Coordinate transformations

lindsayad opened the pull request:
Handle
- rotation
- scaling
- translation in the context of multi-applications

lindsayad committed:
Multiapp coordinate transform

- All existing tests passing with changes in NearestNodeTransfer to
  not use positions, but rather the coordinate transform instead.
- Added rotated, scaled, translated test
- Allow RZ-XYZ interplay; test
- Support 'up_direction' parameter
- Pick the y-axis as the canonical up direction for use in conjunction
  with the 'up_direction' parameter
- Choose to create rotation matrices in the forward sense, e.g. the user
  should supply angles that take the current domain and rotate the domain
  to the canonical domain
- Support manual rotation of RZ geometries in a limited sense
  - E.g. we can maintain a list of "blessed" manual rotation angles
- Fully support rotation of RZ geometries using the 'up_direction' parameter
  - Although note that there are some rotations that result in a negative
    radius, for which we will error

Closes #12293

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/977614/) on ee10207 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20451/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/977622/) on ee10207 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>805902</th>
      <th colspan=4>#20451 ee1020</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.40%</td>
      <td>82.43%</td>
      <td>+0.03%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73382</td>
      <td>73433</td>
      <td>+51</td>
      <td>63</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15669</td>
      <td>15650</td>
      <td>-19</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20451/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20451/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20451/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Python 3.9](https://civet.inl.gov/job/993456/) on 14b580a : invalidated by @lindsayad

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/996277/) on e6cbd04 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20451/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20451/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fa23b08e5ff053feebf7aeebbcbc64f7733f2ebe`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/996494/) on b577c66 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20451/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20451/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fa23b08e5ff053feebf7aeebbcbc64f7733f2ebe`

fdkong reviewed (changes requested):
It looks very good. Easy to use

loganharbour commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -177,6 +178,38 @@ FEProblemBase::validParams()
+  MooseEnum what_is_up("X=0 Y=1 Z=2");
+  params.addParam<MooseEnum>(
+      "what_is_up",
+      what_is_up,
```
![image](https://user-images.githubusercontent.com/22062133/160951699-216c818b-1300-49cf-a112-0fedf87de3e9.png)

loganharbour commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,276 @@
+    }
+#ifndef NDEBUG
+    else
+      mooseAssert(false, "we should never get here");
```
In non-debug, you should get
```
else
  (void)0
```

which the compiler will throw away; so - you shouldn't need the `ifndef`

We should really make an

```
mooseUnreachable()
```

macro or somethin

loganharbour commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,276 @@
+  //
+  // rotation
+  //
+  const bool has_alpha = params.isParamValid("alpha_rotation");
```
Maybe put the parameters this guy expects in `MooseCordTransform`?

Something like

```
static InputParameters transformParameters();
```

?

loganharbour commented on framework/include/utils/MooseCoordTransform.h:
```diff
@@ -0,0 +1,79 @@
+   */
+  void setTranslationVector(const libMesh::Point & translation);
+
+  enum Direction : unsigned int
+  {
+    X = 0,
+    Y,
+    Z,
+    INVALID
+  };
+
+  void setDestinationCoordinateSystem(Moose::CoordinateSystemType destination_coord_type,
+                                      Direction rz_r_axis = INVALID,
+                                      Direction rz_z_axis = INVALID);
+
+  Moose::CoordinateSystemType coordinateSystem() const { return _coord_type; }
+  Direction rzRAxis() const { return _rz_r_axis; }
+  Direction rzZAxis() const { return _rz_z_axis; }
+
+  void setupFromParameters(const InputParameters & params);
+  void setRotation(Direction up_direction, Direction rz_z_axis = Y);
+  void setRotation(Real alpha, Real beta, Real gamma);
+  void setRotation(Real alpha, Real beta, Real gamma, Direction rz_z_axis);
+  void setLengthUnitsPerMeter(Real length_units_per_meter);
+  void setCoordinateSystem(Moose::CoordinateSystemType);
+
+private:
+  std::unique_ptr<libMesh::RealTensorValue> _scale;
+  std::unique_ptr<libMesh::RealTensorValue> _rotate;
+  libMesh::Point _translation;
+  Moose::CoordinateSystemType _coord_type = Moose::COORD_XYZ;
+  Direction _rz_r_axis = INVALID;
+  Direction _rz_z_axis = INVALID;
+
+  Moose::CoordinateSystemType _destination_coord_type = Moose::COORD_XYZ;
+  Direction _destination_rz_r_axis = INVALID;
+  Direction _destination_rz_z_axis = INVALID;
+
+  bool _has_different_coord_sys = false;
```
Where are ya doxygen?

loganharbour reviewed (commented):
All minor comments. All in all - looks great. Doxygen comment is probably the one that should be taken care of

GiudGiud commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -177,6 +178,38 @@ FEProblemBase::validParams()
   params.addParam<MooseEnum>(
       "rz_coord_axis", rz_coord_axis, "The rotation axis (X | Y) for axisymetric coordinates");
+  params.addParam<Real>("length_units_per_meter",
+                        "How many mesh length units are in a meter, e.g. if your mesh units are "
```
let's have a default of 1?

GiudGiud replied:
also this should be rangechecked

GiudGiud replied:
actually should we have a non-uniform scaling operation instead of this?
It wont commute with the rotation anymore so we'd have to be pretty clear what's going on

lindsayad replied:
> let's have a default of 1?

I like being able to check `isParamValid`

> actually should we have a non-uniform scaling operation instead of this? It wont commute with the rotation anymore so we'd have to be pretty clear what's going on

How would you do this? Like with a `Function`?

GiudGiud replied:
no as in instead of 1 have it '1 (x-direction scaling) 2 (y-direction scaling) 0.1 (z)'
I cant think of a use case rn though so dc too too much

YaqiWang replied:
Why cannot this parameter a parameter of `MooseMesh` or `MooseApp`? I think this is a fairly important design decision. Also, if I set this parameter for cm, probably it won't work for tensor mechanics, how do you address that?

lindsayad replied:
It could live on `MooseApp` but we do not currently have a block for that in the input file. It could also live on `MooseMesh`, however, I think it makes sense to group these new parameters with coordinate information such as `coord_type` which is already in the `Problem` block

YaqiWang replied:
Sounds a good argument. If it was in MooseApp, it will be set through command-line. I think a module can add an error if it never considers this parameter.

lindsayad replied:
Changed this to use the `MooseUnits` system

GiudGiud commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,276 @@
+
+using namespace libMesh;
+
+MooseCoordTransform::MooseCoordTransform() {}
```
cant you say = default in the header?

lindsayad replied:
If you do, then you have to make sure you include the headers of things held by `unique_ptrs`. That's the only drawback to `= default;`

GiudGiud commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,276 @@
+    ret = (*_scale) * ret;
+
+  // If this shows up in profiling we can make _translation a pointer
+  ret += _translation;
```
since this does not commute with the others we ll have to be clear that it's performed after

GiudGiud commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,276 @@
+    ret = 0;
+    ret(_destination_rz_r_axis) = r;
+    ret(_destination_rz_z_axis) = z;
+  }
```
what about the other cases?
_coord_type == Moose::COORD_XYZ && _destination_coord_type == Moose::COORD_RSPH
_coord_type == Moose::COORD_RZ && _destination_coord_type == Moose::COORD_XYZ
etc

lindsayad replied:
RSPH will require special treatment. However, if you are RZ and the "destination" is XYZ then we don't do anything because we are going to choose the RZ system as the "canonical" coordinate system. You can collapse the three-dimensionality of XYZ into the two-dimensionality of RZ in a unique way but you cannot do the reverse. This should be reflected in the new doxygen I added but I will also include it in MooseDocs

GiudGiud replied:
true true. Well I ll look a the moose docs then when you have it. 
There's also 2D XY -> RZ and 2D XZ -> RZ which could be mentioned

GiudGiud reviewed (changes requested):
This also needs a doc page with examples for the cases we expect to be common
(XYZ - XYZ but shift/scale, RZ to XYZ)

If there is a page for explaining `positions` in MultiApp, could be a good idea to link the two together

lindsayad commented on framework/src/transfers/MultiAppNearestNodeTransfer.C:
```diff
@@ -190,15 +206,14 @@ MultiAppNearestNodeTransfer::execute()
                 // Record a local ID for each quadrature point
                 node_index_map[i_proc][key] = outgoing_qps[i_proc].size();
-                outgoing_qps[i_proc].push_back(*node + _to_positions[i_to]);
+                outgoing_qps[i_proc].push_back(to_transform(*node));
```
Yes, we use positions to set a translation vector on the transform

lindsayad committed:
Create MooseCoordTransform::validParams

lindsayad committed:
Doxygen for MooseCoordTransform

lindsayad committed:
Add RSPHERICAL treatment to MooseCoordTransform

With this addition I decided to make all APIs based off the radial
coordinate as opposed to the axial coordinate

lindsayad committed:
Simplify APIs

- Rename API for setting angles using the up_direction parameter
- Remove the radial or axial axes arguments from the rotation setting
  APIs
- Consolidate the two direct angle setting APIs into one

lindsayad committed:
More encapsulation

`setDestinationCoordinateSystem` now just takes a MooseCoordTransform
argument which allows removal of the `rAxis()` and `zAxis()` getters

lindsayad committed:
Test coordinate collapsing

lindsayad committed:
Documentation

lindsayad commented:
Ok this should be ready for re-review

moosebuild commented:
Job [Nonsparse AD](https://civet.inl.gov/job/1027621/) on b2094fb : invalidated by @lindsayad

cticenhour reviewed (dismissed):
Newsletter entry please, especially with an image or two :) Will leave other comments to those who have already been reviewing this code.

hugary1995 commented:
I am a little bit confused by the design. It seems to me that each `FEProblem` holds exactly one instance of `MooseCoordTransform`, then how does it work when we have multiple coordinate systems in one `FEProblem`. It is then ambiguous which coord system should be the destination (the canonical/reference one).

I would at least error out if the user has
```python
[Problem]
  type = FEProblem
  block = '1 3 5'
  coord_type = 'XYZ RZ XYZ'
[]
```
And as I see it, this use case has to be handled at some point in the future, if not in this PR.

hugary1995 commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,338 @@
+  if (map_it == coord_sys.end())
+    setCoordinateSystem(Moose::COORD_XYZ);
+  else
+    setCoordinateSystem(
+        map_it->second,
+        Direction(static_cast<unsigned int>(int(params.get<MooseEnum>("rz_coord_axis")))));
```
So you always set the _coord_type to be the first coordinate type entry in the `coord_sys` map? But down below in the `setDestinationCoordinateSystem` method, you are checking whether the destination coord transformation `_has_different_coord_sys`.

lindsayad replied:
We check *our* `_has_different_coord_sys` but you're right, I should should check `destination_coord_transform._has_different_coord_sys` as well

GiudGiud commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -56,6 +56,119 @@ calculations that utilize the scalar variable, regardless of how this scalar is
+| RZ | RSPHERICAL | RSPHERICAL |
+
+
+Let's consider an example. The below listing shows coordinate transformation
```
the examples are great

lindsayad commented:
> And as I see it, this use case has to be handled at some point in the future, if not in this PR.

It is difficult for me to get motivated to spend a lot of time handling multiple coordinate systems in a single application/problem since we have exactly two uses of it in our modules (for testing the tensor mechanics actions). None in BISON. One in Griffin for testing an action. Zero in Pronghorn.

Do you see yourself using that capability much? And I do error under conditions where the transformation would be ambiguous due to presence of multiple coordinate systems, although as you highlighted I should check the destination `_has_different_coord_sys` as well

dschwen commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,338 @@
+      "coord_type", coord_types, "Type of the coordinate system per block param");
+  params.addParam<MooseEnum>(
+      "rz_coord_axis", rz_coord_axis, "The rotation axis (X | Y) for axisymetric coordinates");
+  params.addParam<Real>("length_units_per_meter",
+                        "How many mesh length units are in a meter, e.g. if your mesh units are "
+                        "centimeters, then this parameter value should be 100.");
```
```suggestion
  params.addParam<std::string>("length_units",
                        "Length units of the mesh, e.g cm, nm, ft, 5in.");
```

GiudGiud replied:
also link to Unit system the coord transfrom documentation in syntax/transfers page pls

GiudGiud reviewed (dismissed):
looks good. Some suggestions to improve the docs. 

Overall what I really want on top of this is a page that says what transfers actually support in terms of direction swaps. Maybe that is different in each transfer but I think there's a lot of common ground. 

I know we support
XY -> XY flipped/rotated/scaled
XY -> RZ (x->r, y->z)
XY -> RSph
XYZ -> XYZ flipped/rotated/scaled

Not sure:
XYZ->RZ (integrating data on cylinders)
(equivalent to XY(Z) -> RZ (X,Y is r,theta plane, Z is set) )
XYZ -> RSpher (integrating data on spherical)
RZ->XYZ (expanding data on cylinders)
Rsphe -> XYZ (expanding data on spheres)

dschwen commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,338 @@
+  // Scale
+  //
+  if (params.isParamValid("length_units_per_meter"))
+    setLengthUnitsPerMeter(params.get<Real>("length_units_per_meter"));
```
```suggestion
    setLengthUnitsPerMeter(Real(MooseUnits(params.get<Real>("length_units_per_meter"))/MooseUnits("m")));
```

hugary1995 replied:
I would suggest setting this to be ratio between _our_ length unit and the _destination_ length unit.

hugary1995 replied:
Also, storing a publicly accessible member `MooseUnits _length_unit` should satisfy @YaqiWang 's needs.

hugary1995 replied:
The downside is that users may no longer specify arbitrary scaling factors. Perhaps we should take either a `std::string` for MooseUnits or a `Real`.

dschwen replied:
You can always specify arbitrary scaling factors. `3.89in` is a valid unit...

hugary1995 replied:
Fantastic!

lindsayad replied:
> I would suggest setting this to be ratio between our length unit and the destination length unit

wherever possible I would like each application's "transformation" information to be as encapsulated as possible. E.g. I would like it's data (as much as possible) to only reflect itself. Because you may have different ratios for example when transferring to/from/between different sub-application. For the "coordinate collapsing" case, that necessitates holding some destination information, which we've already seen can cause confusion ... so if I can limit the potential sources of confusion, I would like to

hugary1995 replied:
I agree with you. I didn't consider the case of multiple coord transformations across different apps.

lindsayad replied:
@dschwen I'm seeing '20cm' get parsed into
```
(gdb) p _length_unit
$2 = {_factor = 0.01, _shift = 0, _base = std::vector of length 1, capacity 1 = {{
      first = MooseUnits::BaseUnit::METER, second = 1}}}
```
It seems like the '20' prefactor is getting ignored

lindsayad replied:
'20*cm' gets parsed correctly

hugary1995 commented:
I don't use that capability much. I was speaking for the most general audience. I agree with you, I think an error at this point is sufficient, until someone has a strong enough motivation.

lindsayad reviewed (commented):
It's weird that I can't request changes on my own pull request. Per slack discussion, we want to see about moving all transform-type information to the `Mesh` block

lindsayad committed:
Per Gary's suggestion, error when destination has multiple coord types

lindsayad committed:
Have length unit parameter use the MooseUnits system

Additionally provide an API for the mesh length unit. Hopefully
this can help Yaqi

lindsayad committed:
Move coord transform parameters to mesh

lindsayad commented:
These perf graph reporter tests are obnoxious. No information in the error messages

lindsayad committed:
Remove extraneous FEProblem::setCoordSys from gold verify file

lindsayad committed:
Use traditional instead of float division in ClipPlugin

lindsayad committed:
Add length unit getter to MooseMesh for Yaqi

lindsayad commented:
I think this is ready for more review

GiudGiud commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,401 @@
+        _z_axis = processZAxis(Y);
+      }
+      else
+        mooseAssert(false, "we should never get here");
```
might as well mooseError then?

lindsayad replied:
sabotages my code coverage haha

GiudGiud commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,401 @@
+        _z_axis = processZAxis(X);
+      }
+      else
+        mooseAssert(false, "we should never get here");
```
same

GiudGiud reviewed (commented):
lgtm

lindsayad commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -56,6 +56,118 @@ calculations that utilize the scalar variable, regardless of how this scalar is
+| XYZ | RSPHERICAL | RSPHERICAL |
+| RZ | RSPHERICAL | RSPHERICAL |
+
+
```
Technically everything is supported. But I can add a couple notes that when transferring data from XYZ to RZ, there may be multiple points in the XYZ space eligible for, say, nearest node transfer and selection of a point will currently be arbitrary

GiudGiud replied:
sounds good

lindsayad committed:
Address miscellanous review points

Includes
- Add warning about non-unique data transforms
- Change some mooseAsserts to mooseErrors

lindsayad committed:
Try to add more clarity around coordinate collapsing

lindsayad commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -56,6 +56,119 @@ calculations that utilize the scalar variable, regardless of how this scalar is
+| XYZ | RZ | RZ |
+| XYZ | RSPHERICAL | RSPHERICAL |
+| RZ | RSPHERICAL | RSPHERICAL |
+
```
I've tried to add a bit more documentation explaining this

lindsayad commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -56,6 +56,119 @@ calculations that utilize the scalar variable, regardless of how this scalar is
+
+1. Rotation about the z-axis by [!param](/Problem/FEProblem/alpha_rotation) degrees
+2. Rotation about the x-axis by [!param](/Problem/FEProblem/beta_rotation) degrees
+3. Rotation about the z-axis (again) by [!param](/Problem/FEProblem/gamma_rotation) degrees
```
My hope is that the `up_direction` parameter route handles the most common cases

lindsayad commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -56,6 +56,119 @@ calculations that utilize the scalar variable, regardless of how this scalar is
+
+!listing transfers/coord_transform/sub-app.i block=Problem
+
+Here, the user is stating that a -90 degree alpha rotation should be applied to
```
yes

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1093254/) on d9a1578 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 93.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 86.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 91.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 83.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 55.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 89.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `ray_tracing` coverage rate (required: 94.0%)
- Failed to generate `rdg` coverage rate (required: 65.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 86.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

lindsayad commented:
@loganharbour any idea what's going on with this one?

loganharbour commented:
Looks like... coverage ran for a previous event but then when the latest ran it didn't clear out the old coverage? This one _actually_ has me confused. Invalidating the recipes that contribute to modules coverage

moosebuild commented:
Job [Modules](https://civet.inl.gov/job/1093241/) on d9a1578 : invalidated by @loganharbour

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1093254/) on d9a1578 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)

lindsayad committed:
Slightly bump nl_abs_tol for xfem test

moosebuild commented:
Job [End user mac](https://civet.inl.gov/job/1098254/) on c41b71f : invalidated by @lindsayad

GiudGiud reviewed (commented):
looks ready to me. One thing to discuss about ordering of operations

lindsayad commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -56,6 +56,137 @@ calculations that utilize the scalar variable, regardless of how this scalar is
+mesh domain described by the MOOSE `Mesh` block to a reference domain. The sequence of
+transformations applied in the `MooseCoordTransform` class is:
+
+1. rotation
+2. scaling
+3. translation
+4. coordinate collapsing
```
An 3D XYZ application/input whose up-direction is in the z-direction would be a candidate for rotation, particularly when considered in the context of data transfer with another application/input that is in 2D RZ (whose up-direction is likely in the +y direction when considering the `_coords` in `Point`)

GiudGiud commented:
I think there is some bounding box logic in the transfers that will need to be adjusted?

lindsayad commented:
I apply the transformation to the bounding box computations in the nearest node transfer, but admittedly I have not thought critically about the interaction between coordinate transformation and bounding box computations.

lindsayad commented:
I should say that I have only introduced the transform into `MultiappNearestNodeTransfer`. At some point we should introduce it to all the transfers, but aren't we going to be reducing the number of transfers soon hopefully with @fdkong's generic transfer?

GiudGiud commented:
actually i think you caught what I was thinking about so I have no solid grounds here :O

GiudGiud commented:
well not exactly. The shapes of the bounding boxes might be totally modified. Could the min and max be flipped by a rotation for example?

lindsayad commented:
Does that matter? All that matters is whether points are in a box

GiudGiud commented:
min and max are used for contains
https://libmesh.github.io/doxygen/classlibMesh_1_1BoundingBox.html#aabb3e5cbc3b7bb5f7e99ddfacc6fdc08

lindsayad commented:
I'll have to think about this more but I have other things I'm working on so it might be a while

lindsayad committed:
Be more careful about bbox transformation

lindsayad commented:
@GiudGiud I think that the newest commit is the right way to handle bounding box rotations

GiudGiud reviewed (changes requested):
@YaqiWang @vincentlaboure @jortensi  please consider whether the order of operation
- rotation
- scaling
- translation
- coordinate system collapse
will work for you. Possibly by running a reactor case.

In my mind, given that translation is dictated by the position (position_file) in the main app, it would be run first.
But maybe I'm wrong and I dont have a lattice reactor input handy to check.

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -295,9 +332,16 @@ MultiAppTransfer::getFromBoundingBoxes()
+    // be careful in constructing the new min and max (first and second)
+    const auto transformed_first = (*_from_transforms[i])(bbox.first);
+    const auto transformed_second = (*_from_transforms[i])(bbox.second);
+    bbox.first = Point(std::min(transformed_first(0), transformed_second(0)),
+                       std::min(transformed_first(1), transformed_second(1)),
+                       std::min(transformed_first(2), transformed_second(2)));
+    bbox.second = Point(std::max(transformed_first(0), transformed_second(0)),
+                        std::max(transformed_first(1), transformed_second(1)),
+                        std::max(transformed_first(2), transformed_second(2)));
```
oh an alternative would be to take the coordinate transform of all 8 corners of the box then find the new min and max corners

lindsayad replied:
K that's what I've implemented

lindsayad commented:
Based off https://gamedev.stackexchange.com/a/16721 and https://docs.microsoft.com/en-us/dotnet/desktop/winforms/advanced/why-transformation-order-is-significant?view=netframeworkdesktop-4.8 I'm going to switch the order of scaling and rotation (although the two commute for isotropic scaling), but those will still occur before translation

lindsayad committed:
Finalize transformation ordering

moosebuild commented:
Job [End user Apple Silicon](https://civet.inl.gov/job/1109490/) on 210cfeb : invalidated by @lindsayad

lindsayad committed:
Add inverse transformation capability

lindsayad committed:
Apply MooseCoordTransform to MeshFunctionTransfer

lindsayad committed:
Apply MooseCoordTransform to ProjectionTransfer

lindsayad committed:
Update InterpolationTransfer to MooseCoordTransform

lindsayad committed:
Update UserObjectTransfer to MooseCoordTransform

lindsayad committed:
Update VariableValueSamplePostprocessorTransfer to MooseCoordTransform

lindsayad committed:
Add transformedPosition API

lindsayad committed:
Update PostprocessorInterpolationTransfer to MooseCoordTransform

lindsayad committed:
Revert "Update VariableValueSamplePostprocessorTransfer to MooseCoordTransform"

This transfer relies on querying the positions/translations of all sub-apps in
the MultiApp, regardless of whether they are local to a process or not.  If a
sub-app is not local then we have no access to its FEProblemBase and
consequently we do not have access to a coordinate transformation object.
Scaling and rotation do not affect the output of MooseCoordTransform::operator()
when the argument is the origin ... but we perform coordinate collapsing
post-translation. I do not want to add some check that errors out if a MultiApp
has a coordinate system such that coordinate collapsing occurs, in which case
`position` potentially does not match MooseCoordTransform::operator()(Point(0)).

This reverts commit 60afbb887cbe96020ecb39c03daa45bd10ae561e.

lindsayad committed:
Add NearestNodeTransfer test with both apps transformed

This kind of test should be performed for all transfers that wish
to show they've successfully adopted the coordinate transformation
object

lindsayad committed:
Remove unused getLocalEntities method

lindsayad committed:
Necessary bbox changes for both-transformed test

I verified that without the changes in this commit, the both-transformed
test passes in serial and with 2 procs, but fails with 4 procs

lindsayad committed:
Add MeshFunctionTransfer test with both apps transformed

lindsayad committed:
Move _bbox_extend_factor up to MultiAppTransfer

lindsayad committed:
Add ProjectionTransfer test with both apps transformed

lindsayad committed:
Add InterpolationTransfer test with both apps transformed

lindsayad commented:
Ok if this passes tests, this is ready for more review/merge

lindsayad committed:
Try tightening tolerance for random contact test failure

lindsayad committed:
The domain-decomposition method seems more robust for this problem

MUMPS seems to randomly fail

GiudGiud reviewed (commented):
reviewed commit by commit hence some comments might be hidden

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -274,16 +329,85 @@ MultiAppTransfer::getFromMultiAppInfo()
+}
 
 std::vector<BoundingBox>
 MultiAppTransfer::getFromBoundingBoxes()
```
this is only called for nearest node, projection and mesh function.
So the bbox_extend_factor parameter is silently ignored for all the others.
Let's suprress it please

lindsayad replied:
What do you mean by suppress it?

lindsayad replied:
Oh I take your meaning. Since there are only a few classes leveraging those APIs I think that for now I will do the inverse of what you recommend ... I will only add the parameter in the derived classes that use this API

GiudGiud replied:
works for me

GiudGiud commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -1084,3 +1085,9 @@ void
+Point
+MultiApp::transformedPosition(const unsigned int app)
+{
+  return appProblemBase(app).coordTransform()(Point(0));
```
nvmd

lindsayad committed:
Add copy construction/assignment unit testing

lindsayad committed:
Change bbox factor concept

And limit the parameter availability to the only transfers that
use `getFromBoundingBoxes` such that we don't mislead users, e.g.
avoid the situation that a user sets this parameter in their transfer
input and then the parameter is silently ignored

lindsayad committed:
Add clarifying comment around setting of destination coordinate systems

GiudGiud reviewed (changes requested):
1) I'm seeing bad results for copy Transfer in both directions. This is kind of an odd case though, as we re transferring using ids for that one instead of coordinates. We are also breaking the assumptions for the transfer from the get-go. 

Let's have a warning for that one if a coord transform exists?

2) we should be able to output these apps in position with all the transforms applied. This would be really helpful for debugging simulations. 

3) length_unit parameter is not very wieldy.
  length_unit = '2*m'
  length_unit = '2'
both got me weird errors that made no sense to me. This is an issue with MooseUnits though

4) I struggled with MeshFunctionTransfer and could not get 180 degree beta rotations to properly catch that the domains had a total overlapp. It could not find points with a tiny z offset in e-16. Alpha work fine. Scaling works fine on its own. Translations work fine

5) Translation + scaling works fine too! i thought we d run into an issue with the order of operation. I should think more about this

6) In the MeshFunctionTransfer it d be really helpful to have a more complete message as to where it's searching for points when it fails. Is it in main app or subapp mesh? Right now all we have:
```
*** ERROR ***
The following error occurred in the object "to_sub", of type "MultiAppMeshFunctionTransfer".

Point not found! (x,y,z)=(      -2,        2,        0)
```

Let me know which ones we should create an issue for and which ones you would do here

lindsayad commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -0,0 +1,453 @@
+{
+  Point ret(point);
+
+  // inverse translate
+  ret -= _translation;
+
+  // inverse rotate and then inverse scale
+  if (_rs_inverse)
+    ret = (*_rs_inverse) * ret;
+
+  // Finally, coordinate system conversions
+  if ((_coord_type == Moose::COORD_XYZ && (_destination_coord_type == Moose::COORD_RZ ||
+                                           _destination_coord_type == Moose::COORD_RSPHERICAL)) ||
+      (_coord_type == Moose::COORD_RZ && _destination_coord_type == Moose::COORD_RSPHERICAL))
+    mooseError("Coordinate collapsing occurred in going to the reference space. There is no unique "
+               "return mapping");
+
```
I don't know what you mean. The operations *are* flipped. Translation goes first (in forward mode, it goes last). And then we apply `_rs_inverse` which is the inverse of the originally applied RS matrix

GiudGiud replied:
yeah my bad, I misunderstood.

lindsayad commented:
> length_unit parameter is not very wieldy.
> length_unit = '2*m'
> length_unit = '2'
> both got me weird errors that made no sense to me. This is an issue with MooseUnits though

I don't know why you'd get a weird error with '2*m'. I have a test that does '5*m'. '2' does not surprise me that it would have some error ... it doesn't have a unit in there. But yes I agree that if the error messages aren't very helpful, we should make them better. But I'm not going to do that in this PR

lindsayad commented:
> I struggled with MeshFunctionTransfer and could not get 180 degree beta rotations to properly catch that the domains had a total overlapp. It could not find points with a tiny z offset in e-16. Alpha work fine. Scaling works fine on its own. Translations work fine

Did you try expanding the bounding boxes for the failing beta case?

lindsayad commented:
> we should be able to output these apps in position with all the transforms applied. This would be really helpful for debugging simulations.

We don't even output sub-apps right now with their positions applied. I would rather not do this now

GiudGiud commented:
> I don't know why you'd get a weird error with 2*m. I have a test that does 5*m. 2 does not surprise me that it would have some error ... it doesn't have a unit in there. But yes I agree that if the error messages aren't very helpful, we should make them better. But I'm not going to do that in this PR
I cant reproduce the error with `2*m` anymore. Must have been a fluke or maybe some parsing error north of it. Or maybe I wrote `2m` ?

> Did you try expanding the bounding boxes for the failing beta case?
Didnt work. The bounding box is *likely* of size 0 in the z direction. This is kind of an edge case

> We don't even output sub-apps right now with their positions applied. I would rather not do this now
Sounds good. Now that you mention positions only, I remember April asked for this and we didnt have bandwidth at the time.

Going to try something else then I ll approve

lindsayad commented:
If you wrote `2m` then I've found that doesn't work

lindsayad commented:
Can you push a failing beta rotation example/test to this branch? I can see if I can fix it easily

lindsayad commented:
similarly if you already have copy transfer inputs that were giving you bad results, please push that as well to make sure my error catches it

lindsayad committed:
Check for identity transforms for MultiAppCopyTransfer

lindsayad committed:
Add 'reference space' to MeshFunctionTransfer error

GiudGiud commented:
sounds good i ll add two tests and push to the branch

lindsayad committed:
Document what transfers support coordinate transformation

lindsayad committed:
Add newsletter entry

GiudGiud commented on framework/src/transfers/MultiAppFieldTransfer.C:
```diff
@@ -131,6 +132,9 @@ MultiAppFieldTransfer::transfer(FEProblemBase & to_problem, FEProblemBase & from
       mooseError("The meshes must be identical to utilize MultiAppCopyTransfer.");
 
+    if (!to_problem.coordTransform().isIdentity() || !from_problem.coordTransform().isIdentity())
+      mooseError("Non-identity transformations cannot be used with MultiAppCopyTransfer");
```
I'd make it a warning. 
If the meshes are the same there's a high chance the user knows what they are doing and that the node id mappings are correct

lindsayad replied:
I would say that instead we should keep it as an error, but we should only error if there is rotation. Post-rotation I don't see how there's any way a user can be confident that the rotated node numberings match

GiudGiud replied:
well if it's just XY to YZ for example a smart user could know what they're doing and have the same mesh in both apps still

lindsayad replied:
I will eat my hat if a user has rotation in one or more of their apps and they are able to get matching dof object numbering post-rotation

lindsayad replied:
I am a libMesh developer and I would not be able to do it without a lot of thought. That would be one damn smart user

GiudGiud replied:
umm I m thinking just load the exact same mesh using a FileMeshGenerator, then rotate one using the coord transform

lindsayad replied:
the exact same mesh? Then post-rotation, they do not 100% overlap which is the predicate of the copy transfer

GiudGiud replied:
but they have matching IDs so it could be done. Error is fine too, was just advocating for non-developing power users, a small group

lindsayad replied:
I will make it a warning

GiudGiud committed:
Add test for copy transfer with coordinate transform

GiudGiud committed:
Add test for postprocessors transfer with coordinate transform, refs #12293

lindsayad committed:
Warnings in CopyTransfer/VariableValueSamplePostprocessorTransfer

lindsayad commented:
@cticenhour I added a newsletter entry

lindsayad committed:
With ops on bboxes, slightly expand to accomodate floating point precision

lindsayad commented:
I'm glad I added those sweep targets. Those always get me later. With the newest commit, things should be good

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1120102/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120114/) on 84f4da8 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 95.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 86.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 90.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 83.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_conduction` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 88.0%)
- Failed to generate `misc` coverage rate (required: 55.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `peridynamics` coverage rate (required: 78.0%)
- Failed to generate `phase_field` coverage rate (required: 89.0%)
- Failed to generate `porous_flow` coverage rate (required: 96.0%)
- Failed to generate `ray_tracing` coverage rate (required: 94.0%)
- Failed to generate `rdg` coverage rate (required: 65.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 86.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 90.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Modules](https://civet.inl.gov/job/1120101/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1120102/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Modules recover](https://civet.inl.gov/job/1120103/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [LibTorch](https://civet.inl.gov/job/1120100/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120114/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120114/) on 84f4da8 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Framework 1](https://civet.inl.gov/job/1120096/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Framework 2](https://civet.inl.gov/job/1120097/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/1120098/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120114/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120114/) on 84f4da8 wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `framework` coverage rate (required: 81.0%)

moosebuild commented:
Job [Framework 1](https://civet.inl.gov/job/1120096/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Framework 2](https://civet.inl.gov/job/1120097/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Framework increased sparsity](https://civet.inl.gov/job/1120098/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Modules](https://civet.inl.gov/job/1120101/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/1120102/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Modules recover](https://civet.inl.gov/job/1120103/) on 84f4da8 : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1120114/) on 84f4da8 : invalidated by @lindsayad
